File: frontend/js/views/credit_card_payoff.js
// Credit-card minimum-payment trap — months and interest paying only the
// declining minimum vs a fixed payment, via /calc/credit-card-payoff. Live.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const months = (n) => {
    if (n == null) return '—';
    const y = Math.floor(n / 12);
    const rem = n % 12;
    return `${n} (${y}y ${rem}m)`;
};
const VIEW = 'credit-card-payoff';
let lastReport = null;
let lastBody = null;

export async function renderCreditCardPayoff(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.ccpayoff.h1.title">// CREDIT CARD PAYOFF</span></h1>
        <p class="muted small" data-i18n="view.ccpayoff.hint.intro">
            The minimum-payment trap. The minimum due is the greater of a floor and a percent of the
            balance, so it shrinks as you pay down — most of each payment goes to interest, stretching
            a small balance into decades. Compare it to a fixed monthly payment. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.ccpayoff.h2.inputs">The card</h2>
            <form id="ccpayoff-form" class="inline-form">
                <label><span data-i18n="view.ccpayoff.label.balance">Balance ($)</span>
                    <input type="number" step="0.01" min="0" name="balance_usd" value="5000" required></label>
                <label><span data-i18n="view.ccpayoff.label.apr">APR (%)</span>
                    <input type="number" step="0.01" min="0" name="apr_pct" value="22" required></label>
                <label><span data-i18n="view.ccpayoff.label.minpct">Minimum payment (% of balance)</span>
                    <input type="number" step="0.1" min="0" name="min_payment_pct" value="2" required></label>
                <label><span data-i18n="view.ccpayoff.label.floor">Minimum payment floor ($)</span>
                    <input type="number" step="1" min="0" name="min_payment_floor_usd" value="25"></label>
                <label><span data-i18n="view.ccpayoff.label.fixed">Fixed payment to compare ($)</span>
                    <input type="number" step="0.01" min="0" name="fixed_payment_usd" value="200"></label>
            </form>
            <div id="ccpayoff-tools" class="ce-toolbar"></div>
            <button type="button" id="ccpayoff-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="ccpayoff-sens" class="ce-sens"></div>
        </div>
        <div id="ccpayoff-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#ccpayoff-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        return {
            balance_usd: Number(fd.get('balance_usd')) || 0,
            apr_pct: Number(fd.get('apr_pct')) || 0,
            min_payment_pct: Number(fd.get('min_payment_pct')) || 0,
            min_payment_floor_usd: Number(fd.get('min_payment_floor_usd')) || 0,
            fixed_payment_usd: Number(fd.get('fixed_payment_usd')) || 0,
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcCreditCardPayoff(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.ccpayoff.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#ccpayoff-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'credit-card-payoff.csv' });
    mount.querySelector('#ccpayoff-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['months_minimum', r.never_pays_off ? 'never' : r.months_minimum],
        ['total_interest_minimum_usd', r.total_interest_minimum_usd],
        ['total_paid_minimum_usd', r.total_paid_minimum_usd],
        ['months_fixed', r.months_fixed == null ? '' : r.months_fixed],
        ['total_interest_fixed_usd', r.total_interest_fixed_usd == null ? '' : r.total_interest_fixed_usd],
        ['interest_saved_usd', r.interest_saved_usd == null ? '' : r.interest_saved_usd],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#ccpayoff-result');
    const minMonths = r.never_pays_off ? t('view.ccpayoff.never') : months(r.months_minimum);
    // Line chart: months to clear the balance as the fixed payment sweeps from the
    // first minimum up to 5× it (payoff time falls steeply as payment rises).
    const lo = Math.max(r.first_minimum_usd || 25, body.balance_usd * 0.01);
    const xs = enh.linspace(lo, lo * 5, 13);
    const pts = await Promise.all(xs.map(async (pay) => {
        const rr = await api.calcCreditCardPayoff({ ...body, fixed_payment_usd: pay });
        return { x: pay, y: rr && rr.months_fixed != null ? rr.months_fixed : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'payment $', ylabel: 'months' });
    const fixedRows = r.months_fixed == null ? '' : `
        <tr><td data-i18n="view.ccpayoff.row.fixedmonths">Months (fixed)</td><td>${months(r.months_fixed)}</td></tr>
        <tr><td data-i18n="view.ccpayoff.row.fixedint">Interest (fixed)</td><td>${money(r.total_interest_fixed_usd)}</td></tr>
        ${r.interest_saved_usd == null ? '' : `<tr class="pos"><td data-i18n="view.ccpayoff.row.saved">Interest saved</td><td>${money(r.interest_saved_usd)}</td></tr>`}`;
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.ccpayoff.h2.result">The cost</h2>
            <div class="cards">
                <div class="card neg"><div class="label" data-i18n="view.ccpayoff.card.minmonths">Minimum-only payoff</div>
                    <div class="value neg">${minMonths}</div></div>
                <div class="card neg"><div class="label" data-i18n="view.ccpayoff.card.minint">Interest (minimum)</div>
                    <div class="value neg">${money(r.total_interest_minimum_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.ccpayoff.card.firstmin">First minimum</div>
                    <div class="value">${money(r.first_minimum_usd)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.ccpayoff.row.minmonths">Months (minimum only)</td><td>${minMonths}</td></tr>
                    <tr><td data-i18n="view.ccpayoff.row.minint">Total interest (minimum)</td><td>${money(r.total_interest_minimum_usd)}</td></tr>
                    <tr><td data-i18n="view.ccpayoff.row.minpaid">Total paid (minimum)</td><td>${money(r.total_paid_minimum_usd)}</td></tr>
                    ${fixedRows}
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#ccpayoff-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: APR 10% → 30%; y: fixed payment from first-min to 5×. Output: months to payoff (lower better → negate).
    const lo = Math.max(base.min_payment_floor_usd || 25, base.balance_usd * 0.02);
    const xVals = enh.linspace(10, 30, 5);
    const yVals = enh.linspace(lo, lo * 5, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'apr_pct', yKey: 'fixed_payment_usd', xVals, yVals, compute: (b) => api.calcCreditCardPayoff(b), pick: (r) => (r && r.months_fixed != null ? r.months_fixed : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells: cells.map((row) => row.map((v) => (v == null ? null : -v))), fmt: (v) => (v == null ? '—' : String(-v) + 'mo'), xfmt: (v) => v.toFixed(0) + '%', yfmt: (v) => '$' + Math.round(v), xName: t('view.ccpayoff.label.apr') || 'APR', yName: t('view.ccpayoff.label.fixed') || 'Payment' });
}

File: crates/traderview-core/src/credit_card_payoff.rs
//! Credit-card minimum-payment trap — simulates paying only the (declining)
//! minimum versus a fixed payment.
//!
//! The minimum due is `max(floor, percent × balance)`, so it shrinks as the
//! balance falls and most of each payment goes to interest — stretching a small
//! balance into decades. A fixed payment kills it far faster.
//!
//! Month by month: `interest = balance × apr/12`, then the payment less
//! interest reduces principal. If a payment can't cover the interest the debt
//! never amortizes (flagged).

use serde::{Deserialize, Serialize};

const MAX_MONTHS: u32 = 1200; // 100-year cap

#[derive(Debug, Clone, Deserialize)]
pub struct CardInput {
    pub balance_usd: f64,
    pub apr_pct: f64,
    /// Minimum payment as a percent of balance (e.g. 2.0).
    pub min_payment_pct: f64,
    /// Minimum payment dollar floor (e.g. 25.0).
    #[serde(default)]
    pub min_payment_floor_usd: f64,
    /// A fixed monthly payment to compare against (0 to skip).
    #[serde(default)]
    pub fixed_payment_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct CardResult {
    /// The first month's minimum payment (for context).
    pub first_minimum_usd: f64,
    /// Months to pay off at the minimum; `None` if it never amortizes.
    pub months_minimum: Option<u32>,
    /// Total interest paid on the minimum path.
    pub total_interest_minimum_usd: f64,
    /// Total paid (balance + interest) on the minimum path.
    pub total_paid_minimum_usd: f64,
    /// Whether the minimum payment fails to ever pay the card off.
    pub never_pays_off: bool,
    /// Months to pay off at the fixed payment; `None` if not supplied/never.
    pub months_fixed: Option<u32>,
    /// Total interest on the fixed path; `None` if not supplied.
    pub total_interest_fixed_usd: Option<f64>,
    /// Interest saved by the fixed payment vs the minimum; `None` if no fixed.
    pub interest_saved_usd: Option<f64>,
}

/// Returns (months, total_interest, paid_off).
fn simulate<F: Fn(f64) -> f64>(balance: f64, monthly_rate: f64, payment_for: F) -> (u32, f64, bool) {
    let mut bal = balance;
    let mut months = 0u32;
    let mut total_interest = 0.0;
    while bal > 0.005 && months < MAX_MONTHS {
        let interest = bal * monthly_rate;
        let pay = payment_for(bal).min(bal + interest);
        if pay <= interest {
            return (months, total_interest, false); // never amortizes
        }
        bal -= pay - interest;
        total_interest += interest;
        months += 1;
    }
    (months, total_interest, bal <= 0.005)
}

pub fn analyze(input: &CardInput) -> CardResult {
    let mr = input.apr_pct / 100.0 / 12.0;
    let pct = input.min_payment_pct / 100.0;
    let floor = input.min_payment_floor_usd;

    let first_minimum = (pct * input.balance_usd).max(floor);

    let (min_months, min_interest, paid_off) =
        simulate(input.balance_usd, mr, |bal| (pct * bal).max(floor));

    let (months_fixed, interest_fixed, interest_saved) = if input.fixed_payment_usd > 0.0 {
        let (fm, fi, fpaid) = simulate(input.balance_usd, mr, |_| input.fixed_payment_usd);
        if fpaid {
            (
                Some(fm),
                Some(fi),
                if paid_off { Some(min_interest - fi) } else { None },
            )
        } else {
            (None, Some(fi), None)
        }
    } else {
        (None, None, None)
    };

    CardResult {
        first_minimum_usd: first_minimum,
        months_minimum: if paid_off { Some(min_months) } else { None },
        total_interest_minimum_usd: min_interest,
        total_paid_minimum_usd: input.balance_usd + min_interest,
        never_pays_off: !paid_off,
        months_fixed,
        total_interest_fixed_usd: interest_fixed,
        interest_saved_usd: interest_saved,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-2
    }

    fn base() -> CardInput {
        CardInput {
            balance_usd: 5000.0,
            apr_pct: 22.0,
            min_payment_pct: 2.0,
            min_payment_floor_usd: 25.0,
            fixed_payment_usd: 200.0,
        }
    }

    #[test]
    fn first_minimum() {
        // max(25, 2% × 5000) = 100.
        assert!(close(analyze(&base()).first_minimum_usd, 100.0));
    }

    #[test]
    fn minimum_trap_takes_decades() {
        let r = analyze(&base());
        assert_eq!(r.months_minimum, Some(968));
        assert!(close(r.total_interest_minimum_usd, 43419.4861));
        assert!(!r.never_pays_off);
    }

    #[test]
    fn total_paid_minimum() {
        let r = analyze(&base());
        assert!(close(r.total_paid_minimum_usd, 5000.0 + 43419.4861));
    }

    #[test]
    fn fixed_payment_far_faster() {
        let r = analyze(&base());
        assert_eq!(r.months_fixed, Some(34));
        assert!(close(r.total_interest_fixed_usd.unwrap(), 1749.8795));
    }

    #[test]
    fn interest_saved() {
        let r = analyze(&base());
        assert!(close(r.interest_saved_usd.unwrap(), 43419.4861 - 1749.8795));
    }

    #[test]
    fn no_fixed_comparison_when_zero() {
        let mut i = base();
        i.fixed_payment_usd = 0.0;
        let r = analyze(&i);
        assert!(r.months_fixed.is_none());
        assert!(r.interest_saved_usd.is_none());
    }

    #[test]
    fn floor_drives_small_balance() {
        // Small balance: floor $25 dominates 2% × 300 = $6.
        let r = analyze(&CardInput {
            balance_usd: 300.0,
            apr_pct: 22.0,
            min_payment_pct: 2.0,
            min_payment_floor_usd: 25.0,
            fixed_payment_usd: 0.0,
        });
        assert!(close(r.first_minimum_usd, 25.0));
        assert!(!r.never_pays_off);
    }

    #[test]
    fn never_pays_off_when_no_floor_and_low_percent() {
        // 1% minimum on 22% APR with no floor: minimum < interest forever.
        let r = analyze(&CardInput {
            balance_usd: 5000.0,
            apr_pct: 22.0,
            min_payment_pct: 1.0,
            min_payment_floor_usd: 0.0,
            fixed_payment_usd: 0.0,
        });
        assert!(r.never_pays_off);
        assert!(r.months_minimum.is_none());
    }
}
